### ISSUE TITLE
warning 9 enabled by default in dev mode

### ISSUE THREAD

**Issue by avsm** [ocaml/dune#8645]
Xavier Leroy notes in https://github.com/ocaml/ocaml/pull/12201#discussion_r1223137675 that warning 9 being enabled by default in dune is not a good default.  Is there a possibility of adjusting the default build flags to be more in line with the decisions taken by upstream OCaml in this regard?

**Comment by rgrinberg** [comment]
Sure, I see nothing wrong with doing it. It will require users to explicitly bump the dune language in their project file though.

Do you have any other warnings you'd like to change?

**Comment by rgrinberg** [comment]
The upstream defaults disable warnings such as

```
6 labels-omitted
Label omitted in function application.
```

and basically all the unused X warnings:

```
32 unused-value-declaration
Unused value declaration. (since 4.00)
33 unused-open
Unused open statement. (since 4.00)
34 unused-type-declaration
Unused type declaration. (since 4.00)
35 unused-for-index
Unused for-loop index. (since 4.00)
36 unused-ancestor
Unused ancestor variable. (since 4.00)
37 unused-constructor
Unused constructor. (since 4.00)
38 unused-extension
Unused extension constructor. (since 4.00)
39 unused-rec-flag
Unused rec flag. (since 4.00)
40 name-out-of-scope
Constructor or label name used out of scope. (since 4.01)
41 ambiguous-name
Ambiguous constructor or label name. (since 4.01)
42 disambiguated-name
Disambiguated constructor or label name (compatibility warning). (since 4.01)
```

In principle, I don't mind. I just wonder if we're really improving UX by doing this.

**Comment by SkySkimmer** [comment]
Disabling 42 is good though.

**Comment by rgrinberg** [comment]
> Disabling 42 is good though.

Indeed, my copy paste was a little bit too wide. Fixed.

**Comment by kit-ty-kate** [comment]
> In principle, I don't mind. I just wonder if we're really improving UX by doing this.

I think the main thing that is bothering users is less which specific warnings are enabled by default, and more that they are enabled as errors.

As a first step, why not at least remove `-warn-error` and only enable `-warn-error` on a potential `dune lint` command (or `dune build @check`) instead so that the errors are not fatal to increase productivity and leave CI systems to enable them as errors.

I also would personally prefer dune to stick with OCaml's standard set of warnings/errors and make them evolve there instead (e.g. https://github.com/ocaml/ocaml/issues/12475) but if it's not possible, at least removing `-warn-error` is to me a much better user-friendly default compared to now.

**Comment by c-cube** [comment]
Seems like a good  time to advocate for the deprecation warning to _not_ be a warn-error in any profile, again
.

**Comment by avsm** [comment]
@kit-ty-kate wrote:
> I also would personally prefer dune to stick with OCaml's standard set of warnings/errors and make them evolve there instead (e.g. https://github.com/ocaml/ocaml/issues/12475) but if it's not possible, at least removing -warn-error is to me a much better user-friendly default compared to now.

I agree. It's the divergence to upstream that's causing confusion, so a good default is to start there and ensure that changes are done in lockstep with the compiler. Non-default Dune profiles can still be available with the current settings, though.

**Comment by rgrinberg** [comment]
I actually don't mind resetting the set of warnings to OCaml's, but removing `-warn-error` is far more work than it looks. The issue is that removing this flag is that compilation will display the warning only when the file needs to be recompiled. So:

```
$ dune build # this warns
$ dune build # no warning appears because compilation already succeeded
```

This is highly undesirable and is going to bring us back to the world before dune where every project was littered with random warnings.

The correct way to fix this is to remove the `-warn-error` but make compilation fail at the end of the build, and replay all the stderr emitted by the various compilation commands. This is all much more work than just removing a flag.

**Comment by gridbugs** [comment]
> compilation will display the warning only when the file needs to be recompiled

Is that really that bad of a problem? OCaml-LSP/Merlin will still show warnings. And since people seem to just workaround this by manually downgrading all the "unused" errors back to warnings anyway we're kind of already in this world (at least for those people).

I have two different use cases for warnings:
 1. While I'm writing code I often do things that cause warnings _temporarily_ like introducing variables that I haven't used yet and it's frustrating to have to prefix them with `_` just to be able to build/run the code. I copy/paste custom flags to downgrade errors into warnings in every dune project I work on. 
 2. When I'm shipping code (either by PR or releasing a project) I want to make sure it has no warnings _but_ not everyone feels this way (see most C projects) and lots of hobby code never gets released anyway so they don't benefit from the current strict default.

I like the idea of making dune attractive to people who just want to mess around with ocaml for fun and that scenario only sees case 1 above. For people using dune in production and who do need case 2, they will still be spending much more time writing code than shipping it so I claim that case 1 what we should be optimizing for when designing UX.

My proposal would be to change the defaults to match the compiler but introduce a flag that treats warnings as errors similar to C's `-werror` or rust's `--deny warnings`.

**Comment by xavierleroy** [comment]
"Upstream" speaking!  I'm glad to see this point being discussed and  the discussion being constructive so far.

FYI, warning 6 "labels-omitted" is now active by default: https://github.com/ocaml/ocaml/pull/10140 .  I would not be surprised if it becomes an error (or maybe a warning that triggers an error by default) at some point in the future.

My feeling is that the default should be "few warnings and very few warnings as errors", to encourage discovery of the language and quick prototyping.  On the other hand, it is currently rather painful to select a reasonable set of warnings to use for production-quality code, so we "upstream" should have more "presets" and better documentation.  For example, did you know that `-W K` activates all the "unused" warnings?  I did not, until I checked the OCaml manual before writing this reply...

**Comment by SkySkimmer** [comment]
Individual warnings got names a few versions ago, maybe the warning groups could have names too?

**Comment by rgrinberg** [comment]
> Is that really that bad of a problem? OCaml-LSP/Merlin will still show warnings. And since people seem to just workaround this by manually downgrading all the "unused" errors back to warnings anyway we're kind of already in this world (at least for those people).

It's pretty bad. If you want to make sure your project is warnings-free you need to do a clean build. If you have the dune cache enabled, you also need to remember to turn it off for the clean build.

Okay, so I think the plan of action is pretty clear:

1. We revert to OCaml's default set of warnings
2. Introduce a mechanism to set the warning level to dune's current default so people that prefer the old default will have an easy path forward
3. We replace `-warn-error` with something that will replay all the active warnings at the end of every build (without failing).

**Comment by xavierleroy** [comment]
I tried to start an "upstream" discussion here: https://github.com/ocaml/ocaml/discussions/12557 .  Please contribute!

**Comment by ejgallego** [comment]
I do like that dune is pretty strict about warnings in `dev` mode, unaddressed warnings seem pretty common in the OCaml world and are a pain, so IMHO the default setup should be the one that promotes better quality code.

It seems to me that devs take always the quicker path, which means that if Dune's dev profile becomes weaker few devs will invest time in going back to a sane warning dev setup.

I do agree that it is sometimes annoying in rapid prototyping, so how to address this should be documented more prominently.

**Comment by shonfeder** [comment]
There's been some related discussion on [discuss.ocaml](https://discuss.ocaml.org/t/ocaml-real-pain-points/17457/81?u=shonfeder).

**Comment by smolkaj** [comment]
Another [discussion](https://discuss.ocaml.org/t/in-praise-of-ocaml/13567) with many people taking issue with dune's decision to treat `unused-x` warnings as error in dev mode.

**Comment by smolkaj** [comment]
It seems to me that ideally, we'd want to both:
1. Not treat `unused-x` warnings as errors *during* local development, so newcomers have a good experience for rapid prototype out of the box.
2. Do treat `unused-x` warnings as errors *at the end of* local development, when checking in code.

The issue is of course that Dune has no control over what happens in step (2).

**Comment by Alizter** [comment]
@smolkaj It's not really that we have strong opinions on whether warnings or errors should be better for the user. It's just that warnings cannot be replayed after they are built or cached meaning they are useless after the first build.

The solution to this issue would be to come up with a way so that actions that produce output have a way of storing it and dune knows how to replay it when it is asked for.

Such a change will require some modifications to the engine and won't be easy to pull of correctly. FWIW I have a prototype laying around, and it can replay warnings after running `dune build` twice, but it has a few other problems I haven't been able to solve yet.

**Comment by smolkaj** [comment]
Thank you for elaborating, that makes sense.

FWIW: Would you be equally worried about warnings not being replayed if it was standard practice in the OCaml community to run `dune lint` before checking in code and use CI checks to enforce that? Maybe we can attack things from that angle in parallel?

**Comment by yawaramin** [comment]
@Alizter am I missing something, because I am repeatedly getting the warning? I added to my `dune` file: `(flags :standard -warn-error -32)`

And every time I run `dune build` _after making any net new change in any file in the project_ I get:

```
File "bin/main.ml", line 36, characters 4-5:
36 | let y = 1
         ^
Warning 32 [unused-value-declaration]: unused value y.
```

I am seeing the same behaviour in watch mode.

I don't see how in the normal course of development, the warning could be completely lost. It should keep popping up with each new build of a _net new_ change in the project. During development, this should be happening almost constantly. If we can get the LSP and editor support to highlight warnings like Rust: https://user-images.githubusercontent.com/32419308/94973078-64748a80-04d0-11eb-9011-64fb097af746.png - then it should be very easy to surface the warnings during development even without making them errors.

I have to admit when I started writing this comment, I intended to argue against disabling warnings-as-errors! But when I actually tried it out, the argument didn't stand...

**Comment by beajeanm** [comment]
OCaml lsp already supports warnings, this is what I can see in neovim:

<img width="401" height="58" alt="Image" src="https://github.com/user-attachments/assets/b39e6358-cb5a-41c2-a01f-2b239baa3d96" />

While I strongly in favor of not treating warnings as error during the local development, I have to concede that with dune it comes with one surprising behavior. Things like undo are not net new changes, so you'll get a clean build even if you still have unused variables.

**Comment by Alizter** [comment]
> I don't see how in the normal course of development, the warning could be completely lost. It should keep popping up with each new build of a _net new_ change in the project. 

If you put the same unused value in a compilation unit that has less dependencies you will see that the warnings are not replayed. This is because the particular action that invoked `ocamlc` did not need to be rerun or have its targets fetched from cache. So the side-effect of emitting the warning can be observed.

Until a few versions ago, dune did not have the shared cache enabled by default. This meant that deleting `_build/` and rebuilding would replay any warnings. However we have good reasons for keeping the shared cache enabled by default, we believe it to be very useful, so this unfortunately makes replaying warnings even more difficult. Any rule whose targets are available in the shared cache will not re-run the side-effects of OCaml compilation, in this emitting warnings.

Here is a demo:

```
$ dune init project foo
Entering directory '/tmp/tmp.eIFQ0FB7VO/foo'
Success: initialized project component named foo

$ cd foo/

$ dune build
Error: Program ocamlc not found in the tree or in PATH
 (context: default)
-> required by loading the OCaml compiler for context "default"

$ dune pkg lock

$ (adding lib/foo.ml and empty lib/foo.mli, adding (flags :standard -warn-error -32) to lib/dune)

$ dune exec -- bin/main.exe
File "lib/foo.ml", line 1, characters 4-5:
1 | let x = 10
        ^
Warning 32 [unused-value-declaration]: unused value x.
Hello, World!

$ dune exec -- bin/main.exe
Hello, World!

$ dune exec -- bin/main.exe
Hello, World!

$ (changing bin/main.ml)

$ dune exec -- bin/main.exe
Goodbye, World!

```

If I did edit `lib/foo.ml` then I would cause a recompilation and I would see the warning again as you have observed. In a development environment, we believe it would be confusing if we showed some warnings but not others, when in actual reality there could be many more. This is why the *default* is to warn-error, but we still provide an escape-hatch for users who wish for warnings to be non-fatal during development.

> During development, this should be happening almost constantly. If we can get the LSP and editor support to highlight warnings like Rust: [user-images.githubusercontent.com/32419308/94973078-64748a80-04d0-11eb-9011-64fb097af746.png](https://user-images.githubusercontent.com/32419308/94973078-64748a80-04d0-11eb-9011-64fb097af746.png) - then it should be very easy to surface the warnings during development even without making them errors.

`ocamllsp` is able to communicate with `dune` in watch mode and gather any diagnostics (warnings, errors, alerts) and highlight the relevant locations in the editor. The issue is that this will not persist between sessions.

If I have created `file_a.ml` which has a warning I have ignored and I build things in watch mode. I will see the warning the first time. But as soon as watch mode detects changes and restarts, `file_a.ml` will not be re-compiled leading to the warning no longer appearing in `dune`. (I forget if `ocamllsp` persists the warning after the build, but I find that unlikely).

Nevertheless, if I come back to my project with `file_a.ml` and fire up `dune` and `ocamllsp`, I will never see that warning again until somebody does a fresh build.

The only consistent way to replay warnings during development is to pass `--cache disabled` to `dune` when you want to see them. This is because this value invalidates all the digests of the actions we have and causes all the rules to be rerun. Then you will see the warnings as long as you don't have things already built with `--cache disabled` existing in `_build/`.

To summarise and answer the very common question: "Why does dune use `-warn-error`?" The answer is that for *most* users we believe this to be a more sensible default than hiding warnings unpredictably. We also have various escape hatches for people who can live with the substandard experience. We also have a section on this in our FAQ: https://dune.readthedocs.io/en/latest/faq.html#how-to-make-warnings-non-fatal

> I have to admit when I started writing this comment, I intended to argue against disabling warnings-as-errors! But when I actually tried it out, the argument didn't stand...

There is no right or wrong answer here. Our problem with disabling the setting by default is coming from a usability perspective aiming to give users a consistent, reproducible experience when running `dune build`. Whether or not you actually like warnings or not, you are free to opt in or out.

Hopefully someday we will find the time to implement this feature so that we can display warnings consistently as "upstream" intended. Most of our recent effort has been focused on dune building packages, but these kinds of user-facing changes are still on our radar.

**Comment by dbuenzli** [comment]
I think people are talking past the point of the issue here. 

This issue is not about treating warnings as errors or not, that is an orthogonal issue.

This issue is about not enabling warnings that were not enabled upstream in the first place and likely for a good reason: because they are *utterly* annoying during development (more specifically: the various unused warnings for toplevel structure items). These warnings are useful for checking your code base before you do a release and/or on a specific lint build perhaps in your CI, but they are not ergonomic for development.

I find it a bit concerning that this has not been addressed for more than two years (not to mention that this has been mentioned informally much much much before). Nowadays `ocaml.org` is very adamant that `dune` is the thing you should use to program in OCaml. If you are getting in the language and fooling around you can expect these people to be highly tripped out if not put-off that their builds are failing because they are not using a definition while they are toying with the language.

**Comment by rgrinberg** [comment]
There's no more discussion about changing the default set of warnings, I think we've already reached agreement that dune should stop changing the default warning set. We just need a transition story for people that do want dune's existing set of warnings. I propose that we have a named warning set for what is dune currently setting. Once something like that is implemented, we can drop back to the default set of warnings.

**Comment by nojb** [comment]
> Once something like that is implemented, we can drop back to the default set of warnings.

I'll take a shot at something along these lines shortly. This issue receives so much attention that it is worth doing.

**Comment by rgrinberg** [comment]
Thanks. I'd suggest a new pform for the set of warnings. E.g. `%{dune-warnings}`

**Comment by nojb** [comment]
> Thanks. I'd suggest a new pform for the set of warnings. E.g. `%{dune-warnings}`

See #12766

**Comment by shonfeder** [comment]
@nojb this is closed by https://github.com/ocaml/dune/pull/12766, right?

**Comment by nojb** [comment]
> [@nojb](https://github.com/nojb) this is closed by [#12766](https://github.com/ocaml/dune/pull/12766), right?

Yes! Thanks.
